base-files: add syslog output for sysupgrade

feckert opened the pull request:
This P/R adds two changes.

1. On importing a configuration, a logging should also be added as with the an export.
2. The output of a sysupgrade should not only be in the shell session that is currently being executed, but also in the syslog. To better understand when an upgarde has taken place. For this purpose, the v function is extended by a logger call.

Shellsession output:
```
Mon Feb  1 11:23:05 CET 2021 upgrade: Image metadata not present
Mon Feb  1 11:23:05 CET 2021 upgrade: Reading partition table from bootdisk...
Mon Feb  1 11:23:05 CET 2021 upgrade: Extract boot sector from the image
Mon Feb  1 11:23:05 CET 2021 upgrade: Reading partition table from image...
Mon Feb  1 11:23:05 CET 2021 upgrade: Saving config files...
Mon Feb  1 11:23:05 CET 2021 upgrade: Commencing upgrade. Closing all shell sessions.
Connection to 192.168.0.51 closed by remote host.
Connection to 192.168.0.51 closed.
```

Syslog output:
```
Mon Feb  1 11:23:05 2021 user.info sysupgrade: Image metadata not present
Mon Feb  1 11:23:05 2021 user.info sysupgrade: Reading partition table from bootdisk...
Mon Feb  1 11:23:05 2021 user.info sysupgrade: Extract boot sector from the image
Mon Feb  1 11:23:05 2021 user.info sysupgrade: Reading partition table from image...
Mon Feb  1 11:23:05 2021 user.info sysupgrade: Saving config files...
Mon Feb  1 11:23:05 2021 user.info sysupgrade: Commencing upgrade. Closing all shell sessions.
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Image metadata not present
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Reading partition table from bootdisk...
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Extract boot sector from the image
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Reading partition table from image..
```

In this context I noticed that the following lines are send twice after `Commencing upgrade. Closing all shell sessions.` I haven't had a closer look yet. But is it correct?
```
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Image metadata not present
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Reading partition table from bootdisk...
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Extract boot sector from the image
Mon Feb  1 11:23:06 2021 user.info sysupgrade: Reading partition table from image..
```

feckert committed:
base-files: add logging for configuration import

Make sysupgrade backup import more verbose.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
base-files: add syslog logging for v function

The logging output should not only be displayed in the calling shell
session but also in the syslog. A sysupgrade and a configuration
import, export can thus be traced in the syslog.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

adschm commented on package/base-files/files/lib/upgrade/common.sh:
```diff
@@ -71,6 +71,7 @@ _vn() {
 
 v() {
 	_v "$(date) upgrade: $@"
+	logger -p info -t sysupgrade "$@"
```
Looks broken to me if it only works for v(), but not for vn() ...

feckert replied:
This vn pattern is currently used in the following places in openwrt base

[stage2](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/stage2):
https://github.com/openwrt/openwrt/blob/d33cd383ed64f3b387c8e3111c3c0c663763c0ea/package/base-files/files/lib/upgrade/stage2#L77-L102

For this I don't see any problem with your suggestion to write that also to the syslog.

[fwtool](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/fwtool.sh):
https://github.com/openwrt/openwrt/blob/d33cd383ed64f3b387c8e3111c3c0c663763c0ea/package/base-files/files/lib/upgrade/fwtool.sh#L82-L88

There I have more of a problem dealing with it properly.
Because if I do it this way, it only says "Supported devices:" in the log, which is not correct in my opinion.

feckert replied:
@adschm I have now adjusted the vn use in stage2 in the function kill_remaining, so that this now also uses the function v. The output for the supported device does not make any sense from my point of view in the syslog output. However, it already says that the device is not supported because this used the v function.

feckert committed:
base-files: change logging for upgrade on stage2

Remove vn call in favour of v call. This commit serves as preparation
for removing the vn function call.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

adschm commented:
I do understand that one might change v<->vn to shape logging as necessary in a specific case. But I still don't think that's a good solution. It's not obvious that v will write to log and its sister function vn does not. This will be an introduction to introduce errors.

Since I don't see a way how we can extend the desired functionality to vn, the choice are either to remove vn entirely (which IMO is a step backwards) or create a separate new function for the desired "output and logging" functionality. This new function would then only be available for full lines, and by choosing it the user should be aware of the limitations.

feckert commented:
@adschm @yousong
> Since I don't see a way how we can extend the desired functionality to vn, the choice are either to remove vn entirely (which IMO is a step backwards) or create a separate new function for the desired "output and logging" functionality. This new function would then only be available for full lines, and by choosing it the user should be aware of the limitations.

Why can't that be an option to undo the vn function? As far as I can see, this is only used in 2 places and was recently add with the following commits 
* https://github.com/openwrt/openwrt/commit/7863c33ceaf155783861aa434cc81b5cf37f7a15
* https://github.com/openwrt/openwrt/commit/856b288d27249f535fe15daa6471bb9393e94140
* https://github.com/openwrt/openwrt/commit/00fb51f97e549cb998ef5dc5dbbc33dff650d891

The question is why do we need this function at all? I understand the pattern behind it.
But I don't think it's a good idea to add yet an other function for syslog to bypass the vn pattern.

My motivations are:
I would like to see what happens in the log during the sysupgrade over the LuCI.

@yousong This pattern is from you. Why is this necessarry.  What was your intention?

yousong commented:
> @yousong This pattern is from you. Why is this necessarry. What was your intention?

"n" in the function names are for "-n" argument of echo command.  Those functions were added to mimic the original "echo -n" call and with prefix.  Simple as that.  On terminal, I guess it would have an effect of progressing, with each item appearing one after another.

feckert commented:
> > @yousong This pattern is from you. Why is this necessarry. What was your intention?
> 
> "n" in the function names are for "-n" argument of echo command. Those functions were added to mimic the original "echo -n" call and with prefix. Simple as that. On terminal, I guess it would have an effect of progressing, with each item appearing one after another.

@yousong 
I am aware of that. 
I just ask if it's necessary.
Because of your change, I now have a problem when I want to write this in the syslog.
Your change only means that no new line is added to a vn or _vn.
Is that really necessary?
This is currently only used in two places.
Can't we do it differently and just not offer a function that imitates the echo -n?
If I want to write something in the shell output or into the syslog, I simply use v 
I can also buffer the output and then call the function v.
Wouldn't that be better? Then I do not need an additional function.

```
v "Device $device not supported by this image" 
for k in $dev_keys; do 
  json_get_var dev "$k" 
  devices="$devices $dev" 
done 
v "Supported devices: $devices" 
```

yousong commented:
> > > @yousong This pattern is from you. Why is this necessarry. What was your intention?
> > 
> > 
> > "n" in the function names are for "-n" argument of echo command. Those functions were added to mimic the original "echo -n" call and with prefix. Simple as that. On terminal, I guess it would have an effect of progressing, with each item appearing one after another.
> 
> @yousong
> I am aware of that.
> I just ask if it's necessary.
> Because of your change, I now have a problem when I want to write this in the syslog.
> Your change only means that no new line is added to a vn or _vn.
> Is that really necessary?
> This is currently only used in two places.
> Can't we do it differently and just not offer a function that imitates the echo -n?
> If I want to write something in the shell output or into the syslog, I simply use v
> I can also buffer the output and then call the function v.
> Wouldn't that be better? Then I do not need an additional function.

It's a presentational thing.  In my opinion, if the progressing effect is not that obvious (each item spurting out so fast that human eyes thought they appeared all at the same time), or the progressing effect is not that important compared to the benefit of proper syslog output will bring, then I would say go ahead refactor those "n" suffixed funcs out ;)

feckert committed:
base-files: change logging for upgrade on fwtool

Remove vn call in favour of v call. This commit serves as preparation
for removing the v function call.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
base-files: remove unused vn and _vn function

The remaining vn calls have been ported to v.
Therefore, these functions are no longer needed and will be removed.

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert committed:
base-files: redirect kill ouptut for ash, telnetd and dropbear

If one of the programmes is not running, then we see the following
output in the logs.

`killall: telnetd: no process killed`

To ensure that the log is clean, redirect the output to /dev/null

Signed-off-by: Florian Eckert <fe@dev.tdt.de>

feckert commented:
@yousong @adschm  Is there anything else I can do?
The upgrade in the log now looks like this on my APU3.
```
Wed Feb 10 17:05:24 2021 user.info upgrade: Image metadata not present
Wed Feb 10 17:05:24 2021 user.info upgrade: Reading partition table from bootdisk...
Wed Feb 10 17:05:25 2021 user.info upgrade: Extract boot sector from the image
Wed Feb 10 17:05:25 2021 user.info upgrade: Reading partition table from image...
Wed Feb 10 17:05:25 2021 user.info upgrade: Saving config files...
Wed Feb 10 17:05:25 2021 user.info upgrade: Commencing upgrade. Closing all shell sessions.
Wed Feb 10 17:05:26 2021 user.info upgrade: Image metadata not present
Wed Feb 10 17:05:26 2021 user.info upgrade: Reading partition table from bootdisk...
Wed Feb 10 17:05:26 2021 user.info upgrade: Extract boot sector from the image
Wed Feb 10 17:05:26 2021 user.info upgrade: Reading partition table from image...
Watchdog handover: fd=7
- watchdog -
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending TERM to remaining processes ...
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sh (10244)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to lua (10421)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to ntpd (10497)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sleep (10717)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to ubusd (3362)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to askfirst (3398)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to urngd (3447)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to logd (5899)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to rpcd (5951)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sh (6561)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sh (6562)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sh (6563)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sh (6564)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to ip (6674)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to hostapd (6726)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to wpa_supplicant (6727)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to netifd (6794)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to rngd (6844)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to odhcpd (7006)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to tcsd (7219)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to sleep (7310)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to crond (7363)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to snmpd (8199)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to uhttpd (8297)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to xinetd (8377)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to micrond (8626)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to collectd (8821)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to dnsmasq (9190)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to lua (9533)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to lua (9673)
Wed Feb 10 17:05:26 CET 2021 upgrade: Sending signal TERM to lua (9738)

Wed Feb 10 17:05:29 CET 2021 upgrade: Sending KILL to remaining processes ...

Wed Feb 10 17:05:31 CET 2021 upgrade: Switching to ramdisk...
Wed Feb 10 16:05:33 UTC 2021 upgrade: Performing system upgrade...
Wed Feb 10 16:05:33 UTC 2021 upgrade: Reading partition table from bootdisk...
Wed Feb 10 16:05:33 UTC 2021 upgrade: Extract boot sector from the image
Wed Feb 10 16:05:33 UTC 2021 upgrade: Reading partition table from image...
Wed Feb 10 16:05:33 UTC 2021 upgrade: Writing image to /dev/mmcblk0p1...
Wed Feb 10 16:05:36 UTC 2021 upgrade: Writing image to /dev/mmcblk0p2...
Wed Feb 10 16:05:58 UTC 2021 upgrade: Writing image to /dev/mmcblk0p3...
Wed Feb 10 16:06:06 UTC 2021 upgrade: Writing new UUID to /dev/mmcblk0...
Wed Feb 10 16:06:06 UTC 2021 upgrade: Upgrading bootloader on /dev/mmcblk0...
Wed Feb 10 16:06:07 UTC 2021 upgrade: Upgrade completed
Wed Feb 10 16:06:08 UTC 2021 upgrade: Rebooting system...
umount: can't unmount /dev: Resource busy
umount: can't unmount /tmp: Resource busy
[92443.533339] reboot: Restarting system
PCEngines apu3
coreboot build 20170302
4080 MB ECC DRAM

SeaBIOS (version rel-1.10.0.1)

Press F10 key now for boot menu
```

yousong commented on package/base-files/files/lib/upgrade/stage2:
```diff
@@ -95,27 +95,25 @@ kill_remaining() { # [ <signal> [ <loop> ] ]
-killall -9 ash
+pgrep telnetd >/dev/null && killall -9 telnetd
+pgrep dropbear >/dev/null && killall -9 dropbear
+pgrep ash >/dev/null && killall -9 ash
```
Redirecting stderr to /dev/null should suffice.  There exists a race condition between `pgrep` and `killall`

feckert replied:
@yousong I can do that too, of course.
The question that comes to my mind is, why do we make a difference her?
Especially telnet is not installed on my targets anymore.
If we redirect it now, we no longer get to see it when we sort out these packages in the future.
Can we not dispense with this distinction altogether?

yousong replied:
> The question that comes to my mind is, why do we make a difference her?
> Especially telnet is not installed on my targets anymore.
> If we redirect it now, we no longer get to see it when we sort out these packages in the future.
> Can we not dispense with this distinction altogether?

I traced those sigkill lines to https://github.com/openwrt/openwrt/commit/1e876b4191ea0d159ec74087478e1ef30a6bb5d0 in 2014.  @blogic may help here.

I can only guess that the hard termination of these 3 kinds of process are for minimalising possible outside interferences.

feckert replied:
@yousong I have now redirected the output for ash, dropbear and telnetd to /dev/null as you suggested.

yousong reviewed (approved):
LGTM.  Thanks for the efforts

adschm commented:
I picked the first patch for the moment. I hope I will have time to look at everything soon.

aparcar commented:
@adschm ping :)

aparcar commented:
I rebased your commits and drop the onecs @adschm already merged. Thanks for this contribution the log looks much better!
